src/functional/swing_slide_filter.zig

```diff
@@ -0,0 +1,552 @@
+        const slope: f80 = @max(
+            @min(slope_derivate / usizeToF80(den), upper_bound_line.slope),
+            lower_bound_line.slope,
+        ); // Get optimal slope that minimizes the squared error
+
+        try getIntercept(
+            slope,
+            first_timestamp,
+            uncompressed_values[first_timestamp],
+            &current_line,
+        );
+
+        const end_value: f64 = f80Tof64(evaluate(&current_line, current_timestamp - 1));
+        try appendValue(end_value, compressed_values);
+    } else {
+        try appendValue(current_segment.end_value, compressed_values);
+    }
+
+    try appendIndex(current_timestamp, compressed_values);
+}
+
+pub fn decompress(
+    compressed_values: []const u8,
+    decompressed_values: *ArrayList(f64),
+) error{OutOfMemory}!void {
+    // TODO: Check length of uncompressed_values is modulo 32.
+
+    const compressed_lines_and_index = mem.bytesAsSlice(
+        f64,
+        compressed_values,
+    );
+
+    var current_line: Line = Line{ .slope = 0, .intercept = 0 };
+    var current_segment: Segment = Segment{
+        .start_time = 0,
+        .start_value = 0,
+        .end_time = 0,
+        .end_value = 0,
+    };
+
+    var first_timestamp: usize = 0;
+    var index: usize = 0;
+    while (index < compressed_lines_and_index.len) : (index += 3) {
+        current_segment.start_time = first_timestamp;
+        current_segment.start_value = compressed_lines_and_index[index];
+        current_segment.end_value = compressed_lines_and_index[index + 1];
+        current_segment.end_time = f64ToUsize(compressed_lines_and_index[index + 2] - 1);
+
+        if (current_segment.start_time < current_segment.end_time) {
+            try getSwingLine(&current_segment, &current_line, 0.0);
+            try decompressed_values.append(current_segment.start_value);
+            var current_timestamp: usize = current_segment.start_time + 1;
+            while (current_timestamp < current_segment.end_time) : (current_timestamp += 1) {
+                const y: f64 = f80Tof64(evaluate(&current_line, current_timestamp));
+                try decompressed_values.append(y);
+            }
+            try decompressed_values.append(current_segment.end_value);
+            first_timestamp = current_timestamp + 1;
+        } else {
+            try decompressed_values.append(current_segment.start_value);
+            first_timestamp += 1;
+        }
+    }
+}
+
+test "swing-filter single line compress and decompress" {
+    const alloc = testing.allocator;
+    const uncompressed_values = [_]f64{ 1.5, 2.5, 3.5, 4.5, 5.5, 6.5, 7.5 };
+    var compressed_values = ArrayList(u8).init(alloc);
+    defer compressed_values.deinit();
+    var decompressed_values = ArrayList(f64).init(alloc);
+    defer decompressed_values.deinit();
+
+    const error_bound: f32 = 0.0;
+
+    try compress_swing(uncompressed_values[0..], &compressed_values, error_bound);
```

Comment:
For consistent group `compress_swing()` and `decompress()` and make a newline between `decompress()` and `testing.` like is already done for the majority of the tests.

Reply:
Fixed